fix: Remove podman binary detection from podman-env command

elasticdotventures opened the pull request:
Addresses feedback from reviewers by removing podman binary detection and clarifying the Docker client connectivity approach.

## Key Changes

- **Removed podman binary detection**: Eliminates `isPodmanAvailable()` function and `which podman` checks per reviewer feedback
- **Clarified service description**: Now explicitly refers to "Podman Docker-compatible service" for precision
- **Removed version specificity**: Changed from "podman 4.9.2" requirement to "recent Podman version with Docker API compatibility" 
- **Focused on Docker client only**: Documentation emphasizes only Docker client required, not podman binary

## Addresses Review Feedback

- ✅ **@afbjorklund**: Removed podman detection, focuses purely on Docker socket connectivity
- ✅ **Version specificity**: Removed specific version requirements as requested
- ✅ **Cleaner approach**: No podman binary dependency, only Docker client against Docker-compatible API

## Before/After

### Before
```bash
# Command would check for podman binary availability
eval $(minikube podman-env)  # Could fail if podman binary not found
```

### After  
```bash
# No podman binary check - only connects Docker client to Podman's Docker-compatible service
eval $(minikube podman-env)
export DOCKER_HOST="tcp://192.168.49.2:2376" 
docker images   # Works with Docker client against Podman service
```

## Foundation for Issue #9229

This implementation provides a clean foundation for the future SSH socket approach described in issue #9229, but keeps the current changes minimal and focused. The SSH socket implementation (#9229) would be a separate, more substantial project that can build on this foundation.

## Testing

- Updated tests to remove podman detection expectations
- All existing Docker client functionality preserved  
- Documentation updated to reflect new approach

This approach is more aligned with issue #9229's vision but represents a focused change that can be merged independently before tackling the larger SSH socket project.

elasticdotventures committed:
dockerEnv vs podmanEnv detection

afbjorklund commented on site/content/en/docs/commands/podman-env.md:
```diff
@@ -4,6 +4,14 @@ description: >
+## Requirements
+
+- **Podman version 4.9.2 or newer is required.**
+- Support for Podman v3 and varlink-based communication has been removed. The `podman-env` command now configures your environment to use the Podman REST API socket, as required by Podman v4+.
```
I don't think that varlink is relevant here, as it was removed already in Podman v2:

https://podman.io/blogs/2020/06/29/podman-v2-announce.html

afbjorklund commented on site/content/en/docs/commands/podman-env.md:
```diff
@@ -4,6 +4,14 @@ description: >
+- Support for Podman v3 and varlink-based communication has been removed. The `podman-env` command now configures your environment to use the Podman REST API socket, as required by Podman v4+.
+
+{{% pageinfo color="warning" %}}
+**Note:** If you are using an older version of Podman, please upgrade to at least v4.9.2 to use `minikube podman-env`. Legacy varlink-based workflows are no longer supported.
```
The version 4.9.2 is quite specific, it might as well say 4.x (which is not supported anyway) and later (5.x)

https://endoflife.date/podman

afbjorklund replied:
I don't think 4.9.3 is random*, but upgrading the client is not an issue.
Some people have problems with _downgrading_, but that's different.

* it is the "last" common v4 version, just like 1.9.3 was the last v1 version.
We could upgrade to v5, but it is not available in every distribution yet.

afbjorklund replied:
There are no features in podman 5 (or even in podman 4, but anyway)
that are required by cri-o, as far as I know? Main issue is client compat.

medyagh reviewed (commented):
can you plz post before/after this PR in the description

afbjorklund reviewed (commented):
I don't think we should look for podman, or suggest to people to use the `podman --remote` client. We should only give a path to a docker socket, for a docker client.

Podman Desktop does **not** use podman, and Podman Compose does **not** default to using podman-compose. Instead, they use the Docker API and docker-compose.

Using the podman-remote client, is what is causing these problems with using cri-o.

`Error: unable to connect to Podman socket: server API version is too old. Client "4.0.0" server "3.4.4"`

It is better to only use `docker` as a legacy client,  and `minikube image` for the rest.

That is why it is called "compatibility socket", while the libpod client/connection keeps breaking the backwards API...

elasticdotventures committed:
Remove podman detection from docker-env per review feedback

- Remove DockerBackendDetector and DetectDockerBackend function
- Remove podman-specific logic from dockerEnvVars
- Simplify docker-env to only handle Docker API compatibility

elasticdotventures committed:
Refactor podman-env to use Docker API compatibility per review feedback

- Change podman-env to use Docker client against Podman's Docker-compatible socket
- Remove SSH-based connectivity (Docker client doesn't support SSH keys)
- Simplify podman-env to use standard Docker environment variables
- Update tests to match new Docker API compatibility approach
- Update documentation to clarify Docker client usage
- Remove podman-specific test cases from docker-env_test.go

This addresses the core review feedback about API compatibility issues
and provides a cleaner separation between docker-env and podman-env.

elasticdotventures committed:
Fix whitespace in docker-env files

Remove unnecessary blank lines that were accidentally introduced
when removing podman detection logic.

elasticdotventures committed:
Add Docker runtime support to podman-env command

Extend podman-env compatibility to support both crio and docker container
runtimes, providing users with more deployment flexibility while maintaining
the core Docker API compatibility approach that eliminates API version
conflicts.

This change allows podman-env to work with:
- Podman driver + crio runtime (original support)
- Podman driver + docker runtime (new support)

Both configurations use the same Docker API compatibility approach that
was implemented to address reviewer feedback about SSH-based connectivity
issues and API version mismatches.

elasticdotventures commented:
## Update: Added Docker Runtime Support

I've added an enhancement that extends podman-env compatibility to support both **crio** and **docker** container runtimes, providing users with more deployment flexibility.

### Changes in Latest Commit

- Extended runtime compatibility check to allow both `constants.CRIO` and `constants.Docker`
- Updated error message to reflect support for both runtimes
- Maintains the same Docker API compatibility approach for both runtime types

### New Supported Configurations

1. **Podman driver + crio runtime** (original support)
   ```bash
   minikube start --driver=podman --container-runtime=crio
   eval $(minikube podman-env)
   ```

2. **Podman driver + docker runtime** (new support)  
   ```bash
   minikube start --driver=podman --container-runtime=docker
   eval $(minikube podman-env)
   ```

Both configurations use the same Docker API compatibility approach that eliminates the SSH-based connectivity issues and API version mismatches.

### Testing

Verified that both runtime configurations work correctly:
- Docker client connects successfully via podman-env
- `docker version` reports minikube's Docker daemon version (28.3.3)
- `docker ps` lists containers running in the cluster
- No API version compatibility errors

This change provides more flexibility while preserving the core fix that addresses the reviewer feedback about API compatibility issues.

elasticdotventures committed:
Update podman-env documentation with local development example

Add practical example showing how to build images directly in minikube
and deploy them to Kubernetes without needing a separate registry.
This demonstrates the key value proposition of the Docker API
compatibility approach.

elasticdotventures commented:
@afbjorklund I just wanted to you a line and say thanks for really great advice, I took it all in and I worked up this solution which I'm now running on my local system successfully!   You deserve all the credit.

I also had to submit this, a separate issue in CRIO (my attempt to isolate the reviews)
https://github.com/kubernetes/minikube/pull/21252

elasticdotventures committed:
refactor: address reviewer feedback by removing podman binary detection

elasticdotventures commented:
@afbjorklund Thanks for the feedback. I've updated the PR to remove podman binary detection and version specificity as requested.

Issue #9229 is separate work that shouldn't be muddled with this change. Hopefully this cleaner foundation makes that future SSH socket implementation a bit smaller.

elasticdotventures commented:
😖 still blocked (merge this will fix #21252) .  unsure how to proceed @afbjorklund @medyagh 🙏
